docs/development/design/flyout-redesign.rst

```diff
@@ -0,0 +1,165 @@
+===============================
+Flyout Redesign Design Document
+===============================
+
+.. warning::
+
+    This document details designs that are a **work in progress**.
+
+
+This document describes the design of a new "flyout API"
+to replace our existing :ref:`"footer_html" <api/v2:Undocumented resources and endpoints>` endpoint in APIv2.
+The Read the Docs theme uses this API to put an updated version selector
+with active versions into the lower left menu.
+On other themes, this appears as a floating menu in the lower right.
+
+.. contents:: Contents
+   :local:
+   :backlinks: none
+   :depth: 1
+
+
+Issues with the existing API
+------------------------------
+
+.. figure:: ../../_static/images/design-docs/flyout/flyout-expanded.png
+    :align: right
+    :figwidth: 300px
+    :target: ../../_static/images/design-docs/flyout/flyout-expanded.png
+
+    Read the Docs with an expanded flyout menu in the lower left corner
+
+* The largest problem with the existing ``/api/v2/footer_html`` endpoint
+  is that it returns a blob of HTML.
+  This limits the ability to use this data in ways other than to generate our exact flyout.
+  For example, Sphinx themes other than our own cannot consume this data easily
+  and there's no real integration point for MkDocs at all.
+* Due to how the URLs in the version selector are generated,
+  this API endpoint can be fairly expensive in the worst case for projects with many versions.
+  As it stands now, this API accounts for approximately 15% of the time taken on our webservers
+  at about 70ms for an average request (P95 ~= 150ms, P99 ~= 235ms).
+* The current API is a combination of data about a project that contains information
+  on the live versions, translations, sub- and super-projects, and links to various other things
+  like downloads, the repository on the VCS, and builds for the project on RTD itself.
+  Some of this data never changes (links to the project and builds on RTD itself)
+  and are effectively static while some of it could change after the documentation is built (translations, active versions).
+
+
+Overlap with existing API endpoints
+-----------------------------------
```

Comment:
Just a note here that we are using the proxied endpoint of the footer, this `/_/api/v2/footer_html`, this allow us to share the same domain with some advantages:

- avoid being blocked by some extensions
- Allow to cache the response (docs are served from cloudflare)
- Allow to use the auth backends in .com

And, I've been having some ideas around our APIs... here is an extract from https://github.com/readthedocs/readthedocs.org/pull/8039/

>    We should probably make a distinction between our general API that handles Read the Docs resources,
   vs our APIs that expose features (like server side search, footer, and embed, all of them proxied).
   This way we can version each endpoint separately.

I do agree that the footer shares a lot with the projects/ endpoint. Proxying api v3 may be a little of a problem, because it relies heavily on resolving URLs from the dashboard, but in proxito we don't have the url patterns from the main site.

Reply:
This is a very good point. Maybe we don't need all the APIv3 endpoints but we can have an endpoint that just returns the same data as the APIv3 endpoint and shares code.